core/src/test/java/org/apache/iceberg/TestSnapshotManager.java

```diff
@@ -250,4 +248,449 @@ public void testCherryPickOverwrite() {
+        .appendFile(FILE_A)
+        .commit();
+    long snapshotId = table.currentSnapshot().snapshotId();
+    // Test a basic case of creating and then removing a branch and tag
+    table.manageSnapshots()
+        .createTag("tag1", snapshotId)
+        .commit();
+    table.manageSnapshots()
+        .removeTag("tag1")
+        .commit();
+    TableMetadata updated = table.ops().refresh();
+    SnapshotRef expectedTag = updated.ref("tag1");
+    Assert.assertNull(expectedTag);
+
+    // Test chained creating and removal of a tag
+    table.manageSnapshots()
+        .createTag("tag2", snapshotId)
+        .removeTag("tag2")
+        .commit();
+    Assert.assertEquals(updated, table.ops().refresh());
+    Assert.assertNull(updated.ref("tag2"));
+  }
+
+  @Test
+  public void testRemovingNonExistingTagFails() {
+    AssertHelpers.assertThrows("Removing a non-existing tag should fail",
+        IllegalArgumentException.class, "Tag does not exist: non-existing",
+        () -> table.manageSnapshots().removeTag("non-existing").commit());
+  }
+
+  @Test
+  public void testReplaceBranch() {
+    table.newAppend()
+        .appendFile(FILE_A)
+        .set("wap.id", "123")
+        .stageOnly()
+        .commit();
+    Snapshot firstSnapshot = Iterables.getOnlyElement(table.snapshots());
+    table.manageSnapshots().createBranch("branch1", firstSnapshot.snapshotId()).commit();
+    table.newAppend()
+        .appendFile(FILE_B)
+        .set("wap.id", "456")
+        .stageOnly()
+        .commit();
+    Snapshot secondSnapshot = Iterables.get(table.snapshots(), 1);
+    table.manageSnapshots().createBranch("branch2", secondSnapshot.snapshotId()).commit();
+    table.manageSnapshots().replaceBranch("branch1", "branch2").commit();
+    Assert.assertEquals(table.ops().refresh().ref("branch1").snapshotId(), secondSnapshot.snapshotId());
+  }
+
+  @Test
+  public void testReplaceBranchNonExistingTargetBranchFails() {
+    AssertHelpers.assertThrows("Replacing a non-existing branch should fail",
+        IllegalArgumentException.class, "Target branch does not exist: non-existing",
+        () -> table.manageSnapshots().replaceBranch("non-existing", "other-branch").commit());
+  }
+
+  @Test
+  public void testReplaceBranchNonExistingSourceFails() {
+    table.newAppend()
+        .appendFile(FILE_A)
+        .commit();
+    long snapshotId = table.currentSnapshot().snapshotId();
+    table.manageSnapshots()
+        .createBranch("branch1", snapshotId)
+        .commit();
+    AssertHelpers.assertThrows("Replacing where the source ref does not exist should fail",
+        IllegalArgumentException.class, "Ref does not exist: non-existing",
+        () -> table.manageSnapshots().replaceBranch("branch1", "non-existing").commit());
+  }
+
+  @Test
+  public void testFastForward() {
+    table.newAppend()
+        .appendFile(FILE_A)
+        .commit();
+
+    table.newAppend()
+        .appendFile(FILE_B)
+        .set("wap.id", "123456789")
```

Comment:
This ID isn't really needed. `stageOnly` is sufficient. This is just a way to get Spark to call `stageOnly`.